JS Components: Improve donut meter accessibility

a8ck3n opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Partially address #26049.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->

Some a11y tweaks for the Donut Meter component. Instead of using a hidden TABLE element to describe the meter, we set the role to IMG, add TITLE and DESC elements, and source those via the caller (as props). In this way, the SVG element is directly accessible to screen readers and should be treated correctly.

Removes the need for the TABLE element and any localization as that would be handled by the caller.

#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

Some discussion in #26049.

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1. Run `cd projects/js-packages/storybook && npm run storybook:dev` and confirm that you see DonutMeter component in Storybook. It should be under JS Packages → Components → DonutMeter.
2. Choose the Default story and confirm we have zero violations under the Accessibility tab.
3. Open the Storybook page in Firefox and use the Accessibility inspector to confirm no violations/suggestions there.

![Report](https://user-images.githubusercontent.com/40267301/189291263-eeaccaec-0afe-4030-a53c-8c4c3d8cae48.png)

a8ck3n committed:
Add basic a11y support to SVG element.

- Remove the “aria-hidden” attribute.
- Set the “role” attribute to “img”.
- Add “title” and “description” attributes.
- Make them props on the component.

a8ck3n committed:
Add changelog file.

a8ck3n committed:
Remove the table element.

The table and the localization are no longer needed now that we have title and desc attributes on the SVG element.

Localization should be the responsibility of the caller.

a8ck3n committed:
Merge branch 'trunk' into update/donut-meter-a11y

a8ck3n committed:
Fix linting errors.

a8ck3n committed:
Fix versions.

a8ck3n committed:
Merge branch 'trunk' into update/donut-meter-a11y

a8ck3n committed:
Merge branch 'trunk' into update/donut-meter-a11y

a8ck3n committed:
Merge branch 'trunk' into update/donut-meter-a11y

kangzj commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -40,21 +54,26 @@ const DonutMeter: React.FC< DonutMeterProps > = ( {
 	backgroundColor = '#00BA37', // jetpack green fallback
 	thickness = '3.5',
 	donutWidth = '64px',
+	title = '',
```
What do you think if we do not default title and description? because empty string doesn't make much sense if devs choose to ignore the parameters.

a8ck3n replied:
Yeah, I was wondering about this. Maybe if the values are not set, we could remove the a11y markup? I imagine if we use this as part of a composite component, the a11y fields might be redundant.

I'll have another look at this.

kangzj reviewed (commented):
It works as described 👍 

Left a comment, nothing major.

a8ck3n committed:
Remove default title and desc values.

kangzj commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -26,6 +26,20 @@ export type DonutMeterProps = {
+	 * Not visible. Used for a11y support.
+	 * If not provided, defaults to an empty string.
+	 */
+	title?: string;
```
```suggestion
	title: string;
```

kangzj commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -26,6 +26,20 @@ export type DonutMeterProps = {
+	 * Not visible. Used for a11y support.
+	 * If not provided, defaults to an empty string.
+	 */
+	description?: string;
```
```suggestion
	description: string;
```

kangzj committed:
Update projects/js-packages/components/components/record-meter-donut/index.tsx

kangzj committed:
Update projects/js-packages/components/components/record-meter-donut/index.tsx

kangzj reviewed (dismissed):
LGTM

kangzj commented:
build failed 🤔

a8ck3n committed:
Merge branch 'trunk' into update/donut-meter-a11y

a8ck3n committed:
Make the a11y props optional.

The meter is hidden from screen readers if no values are provided. A11y values are optional as it’s very likely they’d be redundant when part of a larger composition.

a8ck3n commented:
> build failed 🤔

Per discussion in Slack, we've decided to make the a11y props optional. It's possible that they would be redundant when composed in a larger interface. If not provided, we'll hide the SVG from screen readers and assume surrounding markup is providing appropriate labelling.

jsnmoon commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -40,21 +54,29 @@ const DonutMeter: React.FC< DonutMeterProps > = ( {
 	};
 
+	// If we don't have a title or description, hide the meter from screen readers.
+	const isHidden = title.length === 0 && description.length === 0 ? 'true' : 'false';
```
This line would cause a render-breaking error if `title` or `description` were set to `null`. I assume the TypeScript tooling should prevent such usage, but it might be good to add some defensive logic here like so:

```suggestion
	const isHidden = typeof title === 'string' && title.length === 0 && typeof description == 'string' && description.length === 0 ? 'true' : 'false';
```

a8ck3n replied:
Thanks for pointing this out, Jason. I've addressed this in the incoming commits.

a8ck3n committed:
Merge branch 'trunk' into update/donut-meter-a11y

a8ck3n committed:
Guard against potential null values.

Specifically, do not do string processing on null title or description values.

samiff commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -40,21 +54,35 @@ const DonutMeter: React.FC< DonutMeterProps > = ( {
 		return ( segmentCount / totalCount ) * 100;
 	};
 
+	// If we don't have a title or description, hide the meter from screen readers.
+	const isHidden =
+		typeof title === 'string' &&
+		title.length === 0 &&
+		typeof description === 'string' &&
+		description.length === 0
+			? 'true'
+			: 'false';
```
Do you think it'd ever make sense to allow showing the a11y markup if only one of those props is passed rather than needing both?

a8ck3n replied:
Hey @samiff. Thanks for the feedback. That's actually how it works. The markup is only hidden if both strings are empty/null. If one is present, the a11y markup is included. You can test that behaviour via Storybook to confirm.

oskosk committed:
Merge branch 'trunk' into update/donut-meter-a11y

oskosk reviewed (approved):
LGTM

@a8ck3n please check Sami's feedback above

oskosk commented:
r252802-wpcom

a8ck3n commented:
> @a8ck3n please check Sami's feedback above

Thanks for the review/merge @oskosk! I replied to that feedback so we should be good here.
